src/host/tcp.h

```diff
@@ -589,7 +601,15 @@ namespace asynchost
 
     static void on_resolved(uv_getaddrinfo_t* req, int rc, struct addrinfo*)
     {
-      static_cast<TCPImpl*>(req->data)->on_resolved(req, rc);
+      {
+        std::unique_lock<std::mutex> guard(pending_resolve_requests_mtx);
+        pending_resolve_requests.erase(req);
+      }
+
+      if (req->data)
+      {
+        static_cast<TCPImpl*>(req->data)->on_resolved(req, rc);
+      }
```

Comment:
Hmmm. I think you're right about that, we should extend that. In the spirit of being overly cautious I'll make that change, just in case the usage pattern of `TCPImpl` changes.

Reply:
There's a risk that this could lead to recursive locking, if `on_resolve_failed` re-attempts DNS resolution. It doesn't look like any behaviour does at the moment, but I think there's a risk either way.